File: byte_dance/105. 从前序与中序遍历序列构造二叉树.go
package main

// 105. 从前序与中序遍历序列构造二叉树

// 根据一棵树的前序遍历与中序遍历构造二叉树。
//
// 注意:
// 你可以假设树中没有重复的元素。

// func main() {

// }

func buildTree(preorder []int, inorder []int) *TreeNode {
	if preorder == nil || len(preorder) == 0 {
		return nil
	}
	if inorder == nil || len(inorder) == 0 {
		return nil
	}

	root := preorder[0] // 当前根节点的值
	rootIndex := 0      // 根节点在中序遍历中的下标
	for ; rootIndex < len(inorder); rootIndex++ {
		if inorder[rootIndex] == root {
			break
		}
	}

	leftPre := preorder[1 : rootIndex+1] // 左树的前序遍历序列
	rightPre := preorder[rootIndex+1:]   // 右树的前序遍历序列
	leftIn := inorder[:rootIndex]        // 左树的中序遍历序列
	rightIn := inorder[rootIndex+1:]     // 右树的中序遍历序列

	return &TreeNode{
		Val:   root,
		Left:  buildTree(leftPre, leftIn),
		Right: buildTree(rightPre, rightIn),
	}
}

File: C++/105.从前序与中序遍历序列构造二叉树.cpp

// 105. 从前序与中序遍历序列构造二叉树
//
// 根据一棵树的前序遍历与中序遍历构造二叉树。
// 注意:
// 你可以假设树中没有重复的元素。

#include <iostream>
#include <vector>

#include "header.h"

class Solution {
public:
    TreeNode* buildTree(std::vector<int>& preorder, std::vector<int>& inorder) {
        return buildTree(preorder, 0, preorder.size(), inorder, 0, inorder.size());
    }

    TreeNode* buildTree(std::vector<int>& preorder, int preorder_start, int preorder_end, std::vector<int>& inorder, int inorder_start, int inorder_end) {
        // [preorder_start, preorder_end) 左闭右开
        // [inorder_start, inorder_end) 左闭右开
        std::cout << preorder_start << " " << preorder_end << " " << inorder_start << " " << inorder_end << " " << std::endl;
        if (preorder_start >= preorder_end || inorder_start >= inorder_end) {
            std::cout << "跳出" << std::endl;

            return nullptr;
        }

        int root_val = preorder[preorder_start];  // 当前跟节点值
        std::cout << root_val << std::endl;

        int root_index = inorder_start;  // 根结点在中序遍历中的下标
        for (; root_index < inorder_end; root_index++) {
            if (inorder[root_index] == root_val) {
                break;
            }
        }
        int left_length = root_index + 1;
        int right_length = inorder_end - left_length;

        TreeNode* root = new TreeNode(root_val);
        root->left = buildTree(preorder, preorder_start + 1, root_index + 1, inorder, inorder_start, root_index);
        root->right = buildTree(preorder, root_index + 1, preorder_end, inorder, root_index + 1, inorder_end);

        return root;
    }
};

int main() {
    // std::vector<int> preorder = { 3, 9, 20, 15, 7 };
    // std::vector<int> inorder = { 9, 3, 15, 20, 7 };
    std::vector<int> preorder = {1, 2, 3};
    std::vector<int> inorder = {3, 2, 1};
    Solution().buildTree(preorder, inorder);
}
